### PR TITLE
Improve Error Handling and Readibility for downcasting `Decimal128Array`

### PR DESCRIPTION
**Pull request by retikulum**
# Which issue does this PR close?

Part of #3152.

# Rationale for this change

It is very clear in the issue but there are different schemas while downcasting an ArrayRef  to a related arrow array type. This is the new PR of improving downcasting to `Decimal128Array`

# What changes are included in this PR?

- `as_decimal128_array` is created in `datafusion\common\src\cast.rs`
- Refactor parts where following lines are used:
```rust
as_any().downcast_ref::<Decimal128Array>().unwrap()
 ``` 
```rust
as_decimal_array()
 ``` 
- This commit passed related tests.
- Change `get_decimal_value_from_array` functions return type `ScalarValue` to `Result<ScalarValue>`
-  Out of scope but `dyn_cmp_dict` is added to `datafusion\physical-expr\Cargo.toml` for passing `test_dictionary_type_to_array_coersion`. Before adding to this, following error was returned.
```rust 
Error: ArrowError(CastError("Comparing array of type Dictionary(Int32, Utf8) with array of type Dictionary(Int32, Utf8) requires \"dyn_cmp_dict\" feature"))
```
# Are there any user-facing changes?

I am not sure about it.

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -1881,13 +1881,22 @@ impl ScalarValue {
...
         scale: u8,
-    ) -> ScalarValue {
-        let array = array.as_any().downcast_ref::<Decimal128Array>().unwrap();
-        if array.is_null(index) {
-            ScalarValue::Decimal128(None, precision, scale)
-        } else {
-            ScalarValue::Decimal128(Some(array.value(index)), precision, scale)
-        }
+    ) -> Result<ScalarValue> {
+        let array = match as_decimal128_array(array) {
```
Comment (ozankabak) [inline datafusion/common/src/scalar.rs:1885]: Can you not do this with:
```rust
let array = as_decimal128_array(array)?;
if array.is_null(index) {
    Ok(ScalarValue::Decimal128(None, precision, scale))
} else {
    let value = array.value(index);
    Ok(ScalarValue::Decimal128(Some(value), precision, scale))
}
```

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -2073,14 +2082,14 @@ impl ScalarValue {
...
         scale: u8,
-    ) -> bool {
-        let array = array.as_any().downcast_ref::<Decimal128Array>().unwrap();
+    ) -> Result<bool> {
+        let array = as_decimal128_array(array)?;
         if array.precision() != precision || array.scale() != scale {
-            return false;
+            return Ok(false);
         }
         match value {
```
Comment (ozankabak) [inline datafusion/common/src/scalar.rs:2090]: Using `match` expressions with binary types like `Result` or `Option` is typically less idiomatic than using the `if let` expression. With this in mind, I suggest:
```rust
let is_null = array.is_null(index);
if let Some(v) = value {
  Ok(!is_null && array.value(index) == *v)
} else {
  Ok(is_null)
}
```

**Review by ozankabak — COMMENTED** [review]
Thank you for these changes, much appreciated!

I saw two issues with code style/structure and left inline comments. I think your other PRs may have similar issues in them too. It would be great if you can improve code quality by applying these suggestions to those code pieces too.

Short PRs are typically preferable to long PRs, but IMO this is a case where the opposite is true. Since the same kind of work was broken down to many PRs, the suggestions you get in one of those PRs will probably end up being applicable to many of them. Therefore, it sometimes is better to have one cohesive PR in cases like this.

**Review by alamb — APPROVED** [review]
Thank you @retikulum . While I agree with @ozankabak 's suggestions I also think this PR is an improvement over the current state of the code so I would also be happy to merge it in as is.

Please let me know if you would like to continue to improve it or if I should plan to merge as is

**Comment by retikulum** [comment]
> Thank you for these changes, much appreciated!
> 
> I saw two issues with code style/structure and left inline comments. I think your other PRs may have similar issues in them too. It would be great if you can improve code quality by applying these suggestions to those code pieces too.
> 
> Short PRs are typically preferable to long PRs, but IMO this is a case where the opposite is true. Since the same kind of work was broken down to many PRs, the suggestions you get in one of those PRs will probably end up being applicable to many of them. Therefore, it sometimes is better to have one cohesive PR in cases like this.

Thanks for review @ozankabak. These recommendations are great for improving code quality. I will change the suggested pieces. 

However, I was confused with PR sizes. While we discussed implementation in the issue, @alamb also [suggested](https://github.com/apache/arrow-datafusion/issues/3152#issuecomment-1292654025) that these PRs should be small ( but he suggested a few functions unlike my one function per PR). Rather than implementing one function at one PR, it seems better to implement 3-5 (or more) functions for a PR. My next PRs will include more function implementation.

**Comment by retikulum** [comment]
> Thank you @retikulum . While I agree with @ozankabak 's suggestions I also think this PR is an improvement over the current state of the code so I would also be happy to merge it in as is.
> 
> Please let me know if you would like to continue to improve it or if I should plan to merge as is

Thanks for the feedback. I will change suggested parts and resolve conflicts tomorrow then it will be ready to merge.

**Comment by ozankabak** [comment]
> Thanks for review @ozankabak. These recommendations are great for improving code quality. I will change the suggested pieces.
Great, happy to help!

> ... (but he suggested a few functions unlike my one function per PR) ...
This is basically what I meant as well. This kind of grouping makes sense for PRs like this.

**Comment by alamb** [comment]
> However, I was confused with PR sizes. While we discussed implementation in the issue, @alamb also https://github.com/apache/arrow-datafusion/issues/3152#issuecomment-1292654025 that these PRs should be small ( but he suggested a few functions unlike my one function per PR). Rather than implementing one function at one PR, it seems better to implement 3-5 (or more) functions for a PR. My next PRs will include more function implementation.

Yeah there is no specific rule here.

Basically I have found the smaller your PRs the easier it will be to find enough someone with enough contiguous time to review them. For larger PRs (e.g. 500+ lines) I often need to spend 30-60 minutes to properly review them and finding blocks of time like that for me is much harder than finding 5-10 minutes for smaller PRs

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @retikulum

**Pull request merged**